frontend/src/v5/ui/routes/viewer/viewer.tsx

```diff
@@ -15,23 +15,115 @@
  *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
 
+import { useEffect } from 'react';
 import { ViewerGui } from '@/v4/routes/viewerGui';
-import { useParams } from 'react-router-dom';
-import { ViewerParams } from '../routes.constants';
-import { useContainersData } from '../dashboard/projects/containers/containers.hooks';
+import { useHistory, useParams, generatePath } from 'react-router-dom';
+import { formatMessage } from '@/v5/services/intl';
+import { DialogsActionsDispatchers } from '@/v5/services/actionsDispatchers/dialogsActions.dispatchers';
+import { canUploadToBackend } from '@/v5/store/containers/containers.helpers';
+import { PROJECTS_LIST_ROUTE, ViewerParams } from '../routes.constants';
 import { useFederationsData } from '../dashboard/projects/federations/federations.hooks';
+import { useContainersData } from '../dashboard/projects/containers/containers.hooks';
+import { NoRevisionOverlay } from './noRevisionOverlay';
 
 export const Viewer = () => {
 	const { teamspace, containerOrFederation, revision } = useParams<ViewerParams>();
+	const history = useHistory();
 	const v4Match = {
 		params: {
 			model: containerOrFederation,
 			teamspace,
 			revision,
 		} };
 
-	useFederationsData();
-	useContainersData();
+	const { federations } = useFederationsData();
+	const { containers } = useContainersData();
+
+	const getContainerFromId = (containerId: string) => containers.find((container) => container._id === containerId);
+	const getFederationFromId = (federationId: string) => federations.find(
+		(federation) => federation._id === federationId,
+	);
+	const selectedFederation = getFederationFromId(containerOrFederation);
+	const selectedContainer = getContainerFromId(containerOrFederation);
+
+	const checkLatestRevisionReady = (container) => {
```

Comment:
Maybe this whole check (useeffect  logic included) could be shoved into a separate component like `<CheckLatestRevisionReadiness />` .

Reply:
done